File: src/behavioral/strategy.java
@Data
public class CreditCard {
    private int amount = 1_000;
    private final String number;
    private final String date;
    private final String cvv;
}

@Setter
public class PaymentService {
    private int cost;
    private boolean includeDelivery;

    public void processOrder() {
        // Pop-up to collect card details...
        CreditCard card = new CreditCard("cardNumber", "expiryDate", "cvv");
        // Validate credit card...
        System.out.println("Paying " + getTotal() + " using Credit Card");
        card.setAmount(card.getAmount() - getTotal());
    }

    private int getTotal() {
        return includeDelivery ? cost + 10 : cost;
    }
}

public interface PaymentStrategy {
    void pay(int amount);
}

public class PaymentByCreditCard implements PaymentStrategy {
    private CreditCard card;

    @Override
    public void pay(int amount) {
        // Pop-up to collect card details...
        card = new CreditCard("cardNumber", "expiryDate", "cvv");
        // Validate credit card...
        System.out.println("Paying " + amount + " using Credit Card");
        card.setAmount(card.getAmount() - amount);
    }
}

public class PaymentByPayPal implements PaymentStrategy {
    private String email;
    private String password;

    @Override
    public void pay(int amount) {
        // Pop-up to collect PayPal mail and password...
        email = "...";
        password = "...";
        // Validate account...
        System.out.println("Paying " + amount + " using PayPal");
    }
}

File: src/behavioral/strategy.ts
class CreditCard {
  private amount: number = 1000;

  constructor(
    private number: string,
    private date: string,
    private cvv: string,
  ) {
    this.number = number;
    this.date = date;
    this.cvv = cvv;
  }

  getAmount(): number {
    return this.amount;
  }

  setAmount(amount: number): void {
    this.amount = amount;
  }
}

class PaymentService {
  private cost: number = 0;
  private includeDelivery: boolean = false;

  processOrder(): void {
    // Pop-up to collect card details...
    const card = new CreditCard("cardNumber", "expiryDate", "cvv");
    // Validate credit card...
    console.log("Paying " + this.getTotal() + " using Credit Card");
    card.setAmount(card.getAmount() - this.getTotal());
  }

  private getTotal(): number {
    return this.includeDelivery ? this.cost + 10 : this.cost;
  }
}

interface PaymentStrategy {
  pay(amount: number): void;
}

class PaymentByCreditCard implements PaymentStrategy {
  private card!: CreditCard;

  pay(amount: number): void {
    // Pop-up to collect card details...
    this.card = new CreditCard("cardNumber", "expiryDate", "cvv");
    // Validate credit card...
    console.log("Paying " + amount + " using Credit Card");
    this.card.setAmount(this.card.getAmount() - amount);
  }
}

class PaymentByPayPal implements PaymentStrategy {
  private email!: string;
  private password!: string;

  pay(amount: number): void {
    // Pop-up to collect PayPal mail and password...
    this.email = "...";
    this.password = "...";
    // Validate account...
    console.log("Paying " + amount + " using PayPal");
  }
}
